File: lru_and_lfu/lru_cache.go
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2023-01-17 09:56:21
 * @LastEditTime: 2023-01-17 10:53:58
 */
package lruandlfu

/**
手动实现java中的LinkedHashMap
*/
// 双向链表中的节点类型
type Node struct {
	Prev *Node
	Next *Node
	Key  int
	Val  int
}

// 之所以使用双向链表，目的是各种链表操作 比如删除某个节点，其时间复杂度都为O(1)
type DoubleLinkedList struct {
	Size      int
	DummyHead *Node
	DummyTail *Node
}

func InitDoubleLinkedList() *DoubleLinkedList {
	head, tail := &Node{}, &Node{}
	head.Next = tail
	tail.Prev = head

	return &DoubleLinkedList{Size: 0, DummyHead: head, DummyTail: tail}
}

func (dll *DoubleLinkedList) addLast(node *Node) {
	node.Next = dll.DummyTail
	node.Prev = dll.DummyTail.Prev
	dll.DummyTail.Prev.Next = node
	dll.DummyTail.Prev = node
	dll.Size++
}

// 删除链表中的某个节点
func (dll *DoubleLinkedList) remove(node *Node) {
	node.Prev.Next = node.Next
	node.Next.Prev = node.Prev
	// 不要忘记元素个数减1
	dll.Size--
}

func (dll *DoubleLinkedList) removeFirst() *Node {
	if dll.DummyHead.Next == dll.DummyTail {
		return nil
	}

	first := dll.DummyHead.Next
	dll.remove(first)
	return first
}

// 因为题目中还要求O(1)的时间复杂度，所以还需要将双向链表和HashMap对应起来
// 即实现LinkedHashMap，本质上同时维护双向链表和哈希表
type LRUCache struct {
	Cache    *DoubleLinkedList
	Memo     map[int]*Node
	Capacity int
}

func Constructor(capacity int) LRUCache {
	return LRUCache{
		Cache:    InitDoubleLinkedList(),
		Memo:     map[int]*Node{},
		Capacity: capacity,
	}
}

func (this *LRUCache) Get(key int) int {
	if _, ok := this.Memo[key]; !ok {
		return -1
	}

	this.makeKeyFirst(key)
	return this.Memo[key].Val
}

func (this *LRUCache) Put(key int, value int) {
	if _, ok := this.Memo[key]; ok {
		this.Memo[key].Val = value
		this.makeKeyFirst(key)
	} else {
		// 不存在key
		node := &Node{Key: key, Val: value}
		if this.Capacity == this.Cache.Size {
			// 需要删除头部元素
			firstNode := this.Cache.removeFirst()
			delete(this.Memo, firstNode.Key)
		}
		this.Cache.addLast(node)
		this.Memo[key] = node
	}
}

func (this *LRUCache) makeKeyFirst(key int) {
	node := this.Memo[key]

	this.Cache.remove(node)
	this.Cache.addLast(node)
}

File: lru_and_lfu/LRUCache.java
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2023-01-16 13:54:02
 * @LastEditTime: 2023-01-17 10:19:55
 */
package lru_and_lfu;

import java.util.LinkedHashMap;

/**
 * leetcode 146 middle LRU缓存
 * 
 * LRU是一种缓存淘汰策略，全称是Least Recently Used，最近最少使用
 */
public class LRUCache {
    // 之所以使用LinkedHashMap而不是LinkedList，是为了在O(1)的时间内找到对应的key
    // LinkedList中的contains方法是O(n)的
    private LinkedHashMap<Integer, Integer> cache;
    private int capacity;
    
    /**
     * 以正整数capacity作为容量初始化 LRU 缓存
     * @param capacity
     */
    public LRUCache(int capacity) {
        this.cache = new LinkedHashMap<>();
        this.capacity = capacity;
    }
    
    /**
     * 如果关键字key存在于缓存中，则返回关键字的值，否则返回-1 
     * @param key
     * @return
     */
    public int get(int key) {
        if (!cache.containsKey(key)) {
            return -1;
        }

        return makeKeyFirst(key);
    }
    
    private int makeKeyFirst(int key) {
        // 缓存中删除这个key，在队尾添加上这个key
        int val = cache.get(key);
        cache.remove(key);
        cache.put(key, val);
        return val;
    }

    /**
     * 如果关键字key已经存在，则变更其数据值value；
     * 如果不存在，则向缓存中插入该组key-value。
     * 如果插入操作导致关键字数量超过 capacity ，则应该 逐出 最久未使用的关键字。
     * @param key
     * @param value
     */
    public void put(int key, int value) {
        if (cache.containsKey(key)) {
            makeKeyFirst(key);
        } else {
            if (cache.size() == capacity) {
                // 删除头部元素
                Integer firstKey = cache.keySet().iterator().next();
                cache.remove(firstKey);
            }
        }
        cache.put(key, value);
    }
}
